src/Core/Services/Cache/DabCacheService.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Text;
+using System.Text.Json;
+using Azure.DataApiBuilder.Core.Models;
+using Azure.DataApiBuilder.Core.Resolvers;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using ZiggyCreatures.Caching.Fusion;
+
+namespace Azure.DataApiBuilder.Core.Services.Cache;
+
+/// <summary>
+/// Service which wraps an internal cache implementation that enables
+/// provided an in-memory cache for the DAB engine.
+/// </summary>
+public class DabCacheService
+{
+    // Dependencies
+    private readonly IFusionCache _cache;
+    private readonly ILogger? _logger;
+    private readonly IHttpContextAccessor _httpContextAccessor;
+
+    // Constants
+    private const char KEY_DELIMITER = ':';
+
+    /// <summary>
+    /// Create cache service which encapsulates actual caching implementation.
+    /// </summary>
+    /// <param name="cache">Cache implementation.</param>
+    /// <param name="logger">Logger.</param>
+    /// <param name="httpContextAccessor">Accessor which provides httpContext within factory method.</param>
+    public DabCacheService(IFusionCache cache, ILogger<DabCacheService>? logger, IHttpContextAccessor httpContextAccessor)
+    {
+        _cache = cache;
+        _logger = logger;
+        _httpContextAccessor = httpContextAccessor;
+    }
+
+    /// <summary>
+    /// Attempts to fetch response from cache. If there is a cache miss, call the 'factory method' to get a response
+    /// from the backing database.
+    /// </summary>
+    /// <typeparam name="JsonElement">Response payload</typeparam>
+    /// <param name="queryExecutor">Factory method. Only executed after a cache miss.</param>
+    /// <param name="queryMetadata">Metadata used to create a cache key or fetch a response from the database.</param>
+    /// <param name="cacheEntryTtl">Number of seconds the cache entry should be valid before eviction.</param>
+    /// <returns>JSON Response</returns>
+    /// <exception cref="Exception">Throws when the cache-miss factory method execution fails.</exception>
+    public async ValueTask<JsonElement?> GetOrSetAsync<JsonElement>(IQueryExecutor queryExecutor, DatabaseQueryMetadata queryMetadata, int cacheEntryTtl)
+    {
+        string cacheKey = CreateCacheKey(queryMetadata);
+        JsonElement? result = await _cache.GetOrSetAsync(
```

Comment:
For my understanding: Does the `.GetOrSetAsync()` method takes care of creating a cache entry incase of a cache miss?

Reply:
2/3. Working on response, tbd.